src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PENamedTypeSymbol.cs

```diff
@@ -1998,7 +2029,10 @@ bool tryDecodeExtensionType(out bool isExplicit, [NotNullWhen(true)] out TypeSym
                 ParamInfo<TypeSymbol>[] paramInfos = new MetadataDecoder(moduleSymbol, context: this)
                     .GetSignatureForMethod(markerMethod, out signatureHeader, out mrEx);
 
-                if (mrEx is not null || signatureHeader.IsGeneric || paramInfos.Length <= 1)
+                if (mrEx is not null
+                    || signatureHeader.IsGeneric
+                    || paramInfos.Length <= 1
+                    || this.Layout is not { Kind: LayoutKind.Sequential, Alignment: 0, Size: 0 })
```

Comment:
>|| this.Layout is not { Kind: LayoutKind.Sequential, Alignment: 0, Size: 0 }) [](http://example.com/codeflow?start=20&length=77)

It feels like this can be checked even before we call to ```tryDecodeExtensionType```. I think that one looking for this check will have a very hard time finding it because it is unlikely that one would expect to find the check where the  method's signature is verified.

Reply:
I think this location is find for this check. This method does not only concern itself with the method's signature.
I could move it earlier in the method if that helps.

The alternative would be putting the check in `TryGetExtensionInfo`, but then an unexpected layout would result in the type not being considered an extension at all.
The boundary between "not an extension" and a "bad extension" is fungible, but the layout check feels more like the latter category.

Reply:
Oh, I think you're suggesting to do the check in this method, but before the local function is called. That's fine too